### PR TITLE
Feature/pace 57

### PR DESCRIPTION
**Pull request by bvdeenen**
Before this change, interacting with OpendataDiscovery created a flat structure of 1 database/schema/table, so if you called `pace list databases --catalog odd` you'd get all tables in the catalog. This change adds a hierarchy.

```
pace list databases --catalog odd -o json | jq -r '.databases[]|[.id,.display_name,.type]|@csv'
"2","BookShop Data Lake","Data Lake"
"3","BookShop Transactional","Transactional"
"5","User Transactions","Messaging"
"7","KDS Clickstream","Messaging"
"8","Snowflake Sample Data","Samples"

pace list schemas --catalog odd --database 3 -o json | jq -r '.schemas[]|[.id,.name]|@csv'
"schema","BookShop Transactional"

pace list tables --catalog odd --database 3 --schema schema -o json | jq -r '.tables[]|[.id,.name,.schema.name]|@csv'
"15","dim_publishers","BookShop Transactional"
"14","fct_sales","BookShop Transactional"
"13","fct_inventory","BookShop Transactional"
"12","dim_currency","BookShop Transactional"
"11","dim_books","BookShop Transactional"
"10","dim_promo","BookShop Transactional"
"9","customer_tier_sbx","BookShop Transactional"
"8","dim_countries","BookShop Transactional"
"7","dim_cards","BookShop Transactional"
"6","dim_customer","BookShop Transactional"
"5","dim_payment","BookShop Transactional"

pace get data-policy --catalog odd --database 3 --schema schema 14
metadata:
  description: BookShop Transactional
  title: fct_sales
source:
  fields:
  - name_parts:
    - card_uuid
    required: true
    type: varchar
  - name_parts:
    - update_tmst
...

```

### CODE REVIEW

File: app/src/main/kotlin/com/getstrm/pace/catalogs/OpenDataDiscovery.kt [hunk 1]
```diff
@@ -16,82 +16,57 @@ class OpenDataDiscoveryCatalog(configuration: CatalogConfiguration) : DataCatalo
...
-            dataSources.items + getAllDataSources(page + 1, size)
-        } else {
-            dataSources.items
+    /**
+     * We're doing a weird workaround, where we search all dataSources, and then select those
+     * that contain at least one dataset. This is because ODD does not provide information
+     * of this kind in its api.
+     */
+    override suspend fun listDatabases(): List<Database> =
+        dataSources.values.filter{dataSource ->
```
Comment (trietsch) [inline app/src/main/kotlin/com/getstrm/pace/catalogs/OpenDataDiscovery.kt]: Please ensure formatting

File: app/src/main/kotlin/com/getstrm/pace/catalogs/OpenDataDiscovery.kt [hunk 2]
```diff
@@ -16,82 +16,57 @@ class OpenDataDiscoveryCatalog(configuration: CatalogConfiguration) : DataCatalo
...
+            val searchId = searchDataSetsInDataSource(dataSource).searchId
```
Comment (trietsch) [inline app/src/main/kotlin/com/getstrm/pace/catalogs/OpenDataDiscovery.kt:28]: I don't want to make this a blocker now, but have you thought about performance? This is a gql query per datasource

File: app/src/main/kotlin/com/getstrm/pace/catalogs/OpenDataDiscovery.kt [hunk 3]
```diff
@@ -16,82 +16,57 @@ class OpenDataDiscoveryCatalog(configuration: CatalogConfiguration) : DataCatalo
...
-         */
+    class Database(
+        override val catalog: OpenDataDiscoveryCatalog,
+        val dataSource: DataSource,
+        id: String,
+        dbType: String,
+        displayName: String
+    ) : DataCatalog.Database(catalog, id, dbType, displayName) {
+        /*
+            Just returning a hardcoded schema with id 'schema' and name the same as that of the dataSource.
```
Comment (trietsch) [inline app/src/main/kotlin/com/getstrm/pace/catalogs/OpenDataDiscovery.kt:43]: Essentially the cardinality is:
Database <1--1> Schema <1--1..*> Table

Correct?

File: app/src/main/kotlin/com/getstrm/pace/catalogs/OpenDataDiscovery.kt [hunk 4]
```diff
@@ -16,82 +16,57 @@ class OpenDataDiscoveryCatalog(configuration: CatalogConfiguration) : DataCatalo
...
+        private val oddDatabase: Database,
         id: String,
         name: String,
-    ) :
-        DataCatalog.Schema(database, id, name) {
+    ) : DataCatalog.Schema(oddDatabase, id, name) {
         override suspend fun getTables(): List<DataCatalog.Table> {
-            return listOf(Table(catalog, this, id, name))
+            val searchId = catalog.searchDataSetsInDataSource(oddDatabase.dataSource).searchId
+            return catalog.getAllSearchResults(searchId).map { Table(catalog, this, "${it.id}", it.externalName) }
```
Comment (trietsch) [inline app/src/main/kotlin/com/getstrm/pace/catalogs/OpenDataDiscovery.kt:59]: Isn't `it.id` a string?

**Review by trietsch — CHANGES_REQUESTED** [review]

File: app/src/main/kotlin/com/getstrm/pace/catalogs/OpenDataDiscovery.kt (hunk 2 above)
Comment (bvdeenen) [reply app/src/main/kotlin/com/getstrm/pace/catalogs/OpenDataDiscovery.kt:28]: Yes I have thought about it. For now it's a non issue, and it's easy to cache or parallellize. Also, there are not that many data sources, and at least it's far less than the number of tables.

File: app/src/main/kotlin/com/getstrm/pace/catalogs/OpenDataDiscovery.kt (hunk 3 above)
Comment (bvdeenen) [reply app/src/main/kotlin/com/getstrm/pace/catalogs/OpenDataDiscovery.kt:43]: yes

File: app/src/main/kotlin/com/getstrm/pace/catalogs/OpenDataDiscovery.kt (hunk 4 above)
Comment (bvdeenen) [reply app/src/main/kotlin/com/getstrm/pace/catalogs/OpenDataDiscovery.kt:59]: no it's a Long in ODD

**Review by trietsch — APPROVED** [review]

**Pull request merged**